File: internal/platform/jobs/connector_probe.go
package jobs

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"github.com/riverqueue/river"
	"github.com/riverqueue/river/rivertype"

	"github.com/xufei5620/xingmang-platform/internal/platform/connector"
	"github.com/xufei5620/xingmang-platform/internal/platform/ops"
)

// XM-OPS0 connector health probe.
//
// This periodic job answers one question the sync jobs (sub2api_sync /
// newapi_sync) do not: "is the connector itself reachable and running a
// compatible version", independent of whether any business data (users,
// revenue, channels, ...) happened to sync successfully this round. It never
// touches the request path -- the ops overview HTTP query only ever reads
// what this job already wrote to ops.metric_observation, never calls
// Version()/Health() itself.

const (
	// ConnectorProbeJobKind is the stable River kind for this job.
	ConnectorProbeJobKind = "connector_probe"

	// DefaultConnectorProbeInterval matches the Sub2API/NewAPI/FinanceCollect
	// precedent of "5 minutes": frequent enough that the ops page never
	// looks stale, infrequent enough not to hammer either upstream's admin
	// API just to ask "are you there".
	DefaultConnectorProbeInterval = 5 * time.Minute

	// ConnectorProbeStalenessThresholdSeconds is a fixed 900s (15m): three
	// missed 5-minute cycles of slack before the ops page flags a
	// connector's health as stale. Fixed rather than derived from the
	// configured interval, per the ops overview design.
	ConnectorProbeStalenessThresholdSeconds int32 = 900

	// MetricSub2APIConnectorHealth / MetricNewAPIConnectorHealth are the ops
	// metric keys this job writes. Each observation's Value is
	// {"version","supported","healthy","kind","latency_ms","checked_at"}:
	// version/supported come from connector.VersionInfo, healthy/latency_ms/
	// checked_at from connector.HealthResult, and "kind" is
	// HealthResult.ErrorKind (empty string when healthy).
	MetricSub2APIConnectorHealth = "sub2api.connector.health"
	MetricNewAPIConnectorHealth  = "newapi.connector.health"

	connectorProbePrincipalID  = "worker:platform"
	connectorProbeModule       = "platform.jobs"
	connectorProbeMaxAttempts  = 3
	connectorProbeUniquePeriod = DefaultConnectorProbeInterval
)

// ConnectorProbeArgs are the serializable arguments for the probe job.
type ConnectorProbeArgs struct {
	// RunID is reserved for isolated integration tests. Production clients
	// must leave it empty so all replicas share one unique probe record.
	RunID string `json:"run_id,omitempty"`
}

// Kind returns the stable River kind.
func (ConnectorProbeArgs) Kind() string { return ConnectorProbeJobKind }

// InsertOpts gives every probe run the same retry and uniqueness policy.
func (ConnectorProbeArgs) InsertOpts() river.InsertOpts {
	return river.InsertOpts{
		MaxAttempts: connectorProbeMaxAttempts,
		Queue:       QueueMaintenance,
		UniqueOpts: river.UniqueOpts{
			ByArgs:   true,
			ByPeriod: connectorProbeUniquePeriod,
			ByQueue:  true,
			ByState:  rivertype.UniqueOptsByStateDefault(),
		},
	}
}

// probeReadClient is the minimal read surface the probe needs from either
// connector's ReadClientV2. Declared locally instead of importing
// connectors/sub2api or connectors/newapi: both packages' ReadClient
// interfaces already expose exactly these two methods returning
// internal/platform/connector's shared VersionInfo/HealthResult types, so
// any sub2api.ReadClientV2 or newapi.ReadClientV2 value satisfies this
// interface structurally with no explicit conversion or extra dependency.
type probeReadClient interface {
	Version(ctx context.Context) (connector.VersionInfo, error)
	Health(ctx context.Context) (connector.HealthResult, error)
}

// probeClientFactory builds a probeReadClient for one connector round.
//
// This is deliberately narrower than Sub2APIClientFactory/NewAPIClientFactory
// (which return the full sub2api.ReadClientV2/newapi.ReadClientV2 -- many
// methods this worker never calls): the probe only ever calls Version and
// Health, so its dependency surface should say so. Production wiring adapts
// cfg.sub2apiClientFactory()/cfg.newapiClientFactory() into this shape with a
// one-line closure in client.go (any *ReadClientV2 already satisfies
// probeReadClient structurally); tests can inject a minimal stub directly
// without implementing either connector's full interface.
type probeClientFactory func(ctx context.Context) (probeReadClient, error)

// ConnectorProbeOptions are the Worker's construction parameters.
type ConnectorProbeOptions struct {
	Logger      *slog.Logger
	Environment string
	// Store is where probe results land; *ops.Store satisfies it. Reuses
	// the same narrow ObservationStore interface (Get + UpsertWithSample)
	// the sync workers already depend on.
	Store ObservationStore

	// Sub2APISource / NewAPISource become each observation's Source field.
	// Callers should pass the same InstanceID used for the corresponding
	// sync job, so "which connector instance" reads consistently across
	// the sync-pipeline metrics and this health-probe metric.
	Sub2APISource    string
	Sub2APINewClient probeClientFactory
	NewAPISource     string
	NewAPINewClient  probeClientFactory

	// ExpectedInterval documents the configured probe cadence for the
	// observation's rollup metadata (see annotateRollupMetadata). Zero
	// omits that metadata.
	ExpectedInterval time.Duration
	// Now can inject a fixed clock for tests; defaults to time.Now.
	Now func() time.Time
}

// ConnectorProbeWorker periodically calls Version() and Health() on the
// currently effective sub2api and newapi connectors (mode resolved the same
// way the sync jobs resolve it -- see Sub2APINewClient/NewAPINewClient) and
// records the results as ops observations.
type ConnectorProbeWorker struct {
	river.WorkerDefaults[ConnectorProbeArgs]

	logger      *slog.Logger
	environment string
	store       ObservationStore

	sub2apiSource    string
	sub2apiNewClient probeClientFactory
	newapiSource     string
	newapiNewClient  probeClientFactory

	expectedInterval time.Duration
	now              func() time.Time
}

// NewConnectorProbeWorker constructs a worker with safe defaults.
func NewConnectorProbeWorker(opts ConnectorProbeOptions) *ConnectorProbeWorker {
	logger := opts.Logger
	if logger == nil {
		logger = structuredDefaultLogger()
	}
	environment := opts.Environment
	if strings.TrimSpace(environment) == "" {
		environment = "unknown"
	}
	now := opts.Now
	if now == nil {
		now = time.Now
	}
	return &ConnectorProbeWorker{
		logger:           logger,
		environment:      environment,
		store:            opts.Store,
		sub2apiSource:    opts.Sub2APISource,
		sub2apiNewClient: opts.Sub2APINewClient,
		newapiSource:     opts.NewAPISource,
		newapiNewClient:  opts.NewAPINewClient,
		expectedInterval: opts.ExpectedInterval,
		now:              now,
	}
}

// Work probes both connectors and writes one observation each. The two
// probes are independent: a failure probing sub2api never prevents the
// newapi observation (or vice versa) from being attempted or written,
// matching the sync workers' "each metric writes independently" discipline.
//
// The returned error reflects only *store write* failures, never a probed
// connector reporting unhealthy/not_supported/etc -- that outcome is
// recorded honestly as a SyncFailed observation (see failureObservation),
// which is a successful run of this job, not a reason for River to retry.
// Retrying would not fix "production+fake" or "real mode unconfigured", so
// treating those as job failures would just add retry noise.
func (w *ConnectorProbeWorker) Work(ctx context.Context, job *river.Job[ConnectorProbeArgs]) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	now := w.now().UTC()

	observations := [2]ops.Observation{
		w.probe(ctx, now, MetricSub2APIConnectorHealth, w.sub2apiSource, w.sub2apiNewClient),
		w.probe(ctx, now, MetricNewAPIConnectorHealth, w.newapiSource, w.newapiNewClient),
	}

	var writeErr error
	probeOK, probeFailed := 0, 0
	for i := range observations {
		obs := observations[i]
		if obs.Status == ops.SyncOK {
			probeOK++
		} else {
			probeFailed++
		}
		annotateRollupMetadata(&obs, w.expectedInterval)
		if _, err := w.store.UpsertWithSample(ctx, obs); err != nil {
			writeErr = errors.Join(writeErr, fmt.Errorf("write %s: %w", obs.MetricKey, err))
		}
	}

	w.logJob(ctx, job, writeErr == nil, probeOK, probeFailed)
	return writeErr
}

// probe builds a client, calls Version then Health, and returns exactly one
// observation. Version and Health are combined into a single observation
// (rather than two independent metrics) because the required shape is one
// JSON value carrying both, so either read failing -- or the client failing
// to build at all, which covers both "real mode misconfigured" and the
// production+fake guard baked into the dynamic factories (both surface as
// connector.KindNotSupported) -- is treated as the whole probe failing for
// that connector. This means production+fake connectors are simply never
// probed, with no special-case branch needed here: the factory itself
// refuses to hand back a client for that combination.
func (w *ConnectorProbeWorker) probe(
	ctx context.Context, now time.Time, metricKey, source string,
	newClient probeClientFactory,
) ops.Observation {
	base := ops.Observation{
		MetricKey:                 metricKey,
		Source:                    source,
		Environment:               w.environment,
		SyncedAt:                  now,
		StalenessThresholdSeconds: ConnectorProbeStalenessThresholdSeconds,
	}

	client, err := newClient(ctx)
	if err != nil {
		return w.failureObservation(ctx, base, now, connector.KindOf(err))
	}

	versionInfo, verErr := client.Version(ctx)
	if verErr != nil {
		return w.failureObservation(ctx, base, now, connector.KindOf(verErr))
	}
	healthResult, healthErr := client.Health(ctx)
	if healthErr != nil {
		return w.failureObservation(ctx, base, now, connector.KindOf(healthErr))
	}

	base.Status = ops.SyncOK
	base.ObservedAt = &now
	base.LastSuccess = &now
	base.Watermark = versionInfo.Fingerprint
	base.Value = map[string]any{
		"version":    versionInfo.Detected,
		"supported":  versionInfo.Supported,
		"healthy":    healthResult.Healthy,
		"kind":       string(healthResult.ErrorKind),
		"latency_ms": healthResult.LatencyMS,
		"checked_at": healthResult.CheckedAt.UTC().Format(time.RFC3339),
	}
	return base
}

// failureObservation mirrors Sub2APISyncWorker.failureObservation /
// NewAPISyncWorker.failureObservation: it preserves the last-known-good
// Watermark/Value/ObservedAt/LastSuccess when a previous observation exists,
// so a transient probe failure degrades the dashboard's freshness badge
// (via growing staleness) instead of blanking out the last good reading.
func (w *ConnectorProbeWorker) failureObservation(
	ctx context.Context, base ops.Observation, now time.Time, kind connector.ErrorKind,
) ops.Observation {
	if kind == "" {
		kind = connector.KindInternal
	}
	failure := ops.Observation{
		MetricKey:                 base.MetricKey,
		Source:                    base.Source,
		Environment:               base.Environment,
		SyncedAt:                  now,
		Status:                    ops.SyncFailed,
		LastErrorCode:             string(kind),
		StalenessThresholdSeconds: base.StalenessThresholdSeconds,
		Value:                     map[string]any{},
	}
	previous, err := w.store.Get(ctx, base.MetricKey, base.Environment)
	if err != nil {
		return failure
	}
	failure.ObservedAt = previous.ObservedAt
	failure.LastSuccess = previous.LastSuccess
	failure.Watermark = previous.Watermark
	failure.IsPartial = previous.IsPartial
	if len(previous.Value) > 0 {
		failure.Value = previous.Value
	}
	return failure
}

func (w *ConnectorProbeWorker) logJob(
	ctx context.Context, job *river.Job[ConnectorProbeArgs], success bool, probeOK, probeFailed int,
) {
	logger := w.logger
	if logger == nil {
		logger = structuredDefaultLogger()
	}
	jobID, attempt, maxAttempts := int64(0), 1, connectorProbeMaxAttempts
	if job != nil && job.JobRow != nil {
		jobID = job.ID
		if job.Attempt > 0 {
			attempt = job.Attempt
		}
		if job.MaxAttempts > 0 {
			maxAttempts = job.MaxAttempts
		}
	}
	level, errorCode, event := slog.LevelInfo, "", "job_completed"
	if !success {
		level, errorCode, event = slog.LevelError, "connector_probe_write_failed", "job_failed"
	}
	logger.LogAttrs(ctx, level, event,
		slog.String("event", event),
		slog.String("module", connectorProbeModule),
		slog.String("environment", w.environment),
		slog.String("principal_id", connectorProbePrincipalID),
		slog.Int64("job_id", jobID),
		slog.String("job_kind", ConnectorProbeJobKind),
		slog.String("queue", QueueMaintenance),
		slog.Int("attempt", attempt),
		slog.Int("max_attempts", maxAttempts),
		slog.Bool("success", success),
		slog.String("error_code", errorCode),
		// These count connectors that were successfully *probed* (a client
		// built and both reads succeeded), not connectors reporting
		// healthy=true -- an unhealthy-but-successfully-probed connector is
		// still probeOK here; its Value.healthy=false is the honest signal,
		// visible in the observation itself, not summarized into this log.
		slog.Int("connectors_probe_ok", probeOK),
		slog.Int("connectors_probe_failed", probeFailed),
	)
}

File: web/apps/admin-web/src/lib/connectorProbe.ts
import type { BadgeTone } from "@xingmang/ui-primitives";
import type { MetricHistoryItem } from "../api/platform";
import { readConnectorHealthValue } from "./ops";

/** XM-CREDS-TAB-PROBE：「连接与凭据」页签的连接器探测。
 *
 *  数据来自 `ConnectorProbeWorker` 每次探测写下的观测
 *  （`internal/platform/jobs/connector_probe.go`），value 形状是
 *  {version, supported, healthy, kind, latency_ms, checked_at}。
 *
 *  这里只做纯派生，不碰 React：一行探测该显示什么，是可以脱离渲染回答的问题。 */

/** 连接器探测写入的指标键。平台标识不认识时返回空串——调用方据此不发查询，
 *  而不是拼一个后端一定查不到的键然后显示"没有数据"。 */
export function connectorHealthMetricKey(platform: string): string {
  if (platform === "sub2api") return "sub2api.connector.health";
  if (platform === "newapi") return "newapi.connector.health";
  return "";
}

export interface UpstreamSupport {
  label: string;
  tone: BadgeTone;
  /** 一句话解释这个判定意味着什么；不确定时说"不确定"，不猜。 */
  detail: string;
}

/** 把 supported 翻成人话。
 *
 *  这不是一个装饰性徽章：`supported` 是**采集链路的判据**，连接器判假时
 *  这个上游的采集就停。所以判假要说清原因——矩阵没声明支持这个版本，
 *  与"上游坏了"是两件完全不同的事，处理方式也不同（前者改代码里的矩阵，
 *  后者查上游）。 */
export function describeUpstreamSupport(supported: boolean | null, version: string): UpstreamSupport {
  if (supported === true) {
    return {
      label: "矩阵已声明支持",
      tone: "success",
      detail: "连接器的兼容矩阵覆盖了这个版本，采集按已核对过的字段形状进行。",
    };
  }
  if (supported === false) {
    return {
      label: "矩阵未声明支持",
      tone: "warning",
      detail: version
        ? `连接器的兼容矩阵没有覆盖 ${version}。这不等于上游坏了：多半是上游升级后矩阵还没跟上（需要先逐字段核对源码，再加矩阵条目）。在此期间采集判据为假。`
        : "连接器的兼容矩阵没有覆盖这次探测到的版本。在此期间采集判据为假。",
    };
  }
  return {
    label: "未知",
    tone: "neutral",
    detail: "这次探测没有给出兼容矩阵判定（旧样本或探测未完成）。不知道不等于不支持。",
  };
}

/** 探测历史表的一行。纯数据，展示留给组件。 */
export interface ConnectorProbeRow {
  id: string;
  /** 探测执行时刻，优先用 value.checked_at，退回观测/落库时刻。 */
  checkedAt: string;
  /** checkedAt 取自哪个字段，显示成小字——两者相差很大时人得看得见。 */
  checkedAtSource: "探测时刻" | "观测时刻" | "落库时刻";
  version: string;
  supported: boolean | null;
  healthy: boolean | null;
  /** 连接器的错误类别（HealthResult.ErrorKind），健康时为空串。 */
  kind: string;
  latencyMs: number | null;
  /** 观测本身的状态（ok / failed …），与 healthy 不是一回事：
   *  前者是"这次采集有没有跑成功"，后者是"上游健不健康"。 */
  status: string;
  lastErrorCode: string;
  source: string;
}

/** 把 `/metrics/history` 的样本翻成探测行，**最近的在前**。
 *
 *  后端按 observed_at 升序返回；表格里人先看最近一次，所以这里倒过来。 */
export function buildProbeRows(items: MetricHistoryItem[]): ConnectorProbeRow[] {
  return items
    .map((item, index) => {
      const value = readConnectorHealthValue(item.value ?? {});
      const checkedAt = value.checkedAt || item.observed_at || item.synced_at;
      const checkedAtSource: ConnectorProbeRow["checkedAtSource"] = value.checkedAt
        ? "探测时刻"
        : item.observed_at
          ? "观测时刻"
          : "落库时刻";
      return {
        id: `${item.synced_at}#${index}`,
        checkedAt,
        checkedAtSource,
        version: value.version,
        supported: value.supported,
        healthy: value.healthy,
        kind: value.kind,
        latencyMs: value.latencyMs,
        status: item.status,
        lastErrorCode: item.last_error_code,
        source: item.source,
      };
    })
    .reverse();
}

/** 最近一次探测；没有样本时返回 null（调用方据此显示"还没探测过"，
 *  而不是显示一行全是破折号的假数据）。 */
export function latestProbe(rows: ConnectorProbeRow[]): ConnectorProbeRow | null {
  return rows[0] ?? null;
}

/** 延迟的显示值。0 是合法值（本地 fake 连接器就是 0），所以不能用真值判断。 */
export function formatLatency(latencyMs: number | null): string {
  return latencyMs === null ? "—" : `${latencyMs} ms`;
}

/** 健康列的显示值与色调。null 是"这次样本没给"，不染成红色。 */
export function describeProbeHealth(healthy: boolean | null, kind: string): {
  label: string;
  tone: BadgeTone;
} {
  if (healthy === true) return { label: "健康", tone: "success" };
  if (healthy === false) return { label: kind ? `不健康 · ${kind}` : "不健康", tone: "danger" };
  return { label: "未知", tone: "neutral" };
}
